File: leetcode/ImplementTrie.java
class Node{
    private boolean isWord;
    private Map<Character, Node> children = new HashMap<>();

    public void setEndOfWord(boolean isWord)  {
        this.isWord = isWord;
    }

    public boolean isEndOfWord() {
        return isWord;
    }

    public Map<Character, Node> getChildren() {
        return children;
    }
}

class Trie {

    private Node root;

    public Trie() {
        root = new Node();

    }

    public void insert(String word) {
        Node current = root;

        for (char c: word.toCharArray()) {
            current = current.getChildren().computeIfAbsent(c, x -> new Node());
        }
        current.setEndOfWord(true);
    }

    public boolean search(String word) {
        Node current = root;

        for (char c : word.toCharArray() ) {
            current = current.getChildren().get(c);

            if (current == null){
                return false;
            }

        }

        return current.isEndOfWord();
    }

    public boolean startsWith(String prefix) {
        Node current = root;

        for (char c : prefix.toCharArray() ) {
            current = current.getChildren().get(c);

            if (current == null){
                return false;
            }

        }

        return true;

    }
}

File: leetcode/implement_trie.py
from dataclasses import dataclass, field

@dataclass
class Node:

    __children : dict = field(default_factory=lambda: defaultdict(list))
    __end_word : bool = False

    @property
    def children(self):
        return self.__children

    @property
    def end_word(self) -> bool:
        return self.__end_word

    @end_word.setter
    def end_word(self, end_word) -> None:
        self.__end_word = end_word


class Trie:

    def __init__(self):
        self.__root = Node()


    def insert(self, word: str) -> None:

        current = self.__root

        for c in word:
            children = current.children

            if c not in children:
                children[c] = Node()

            current = children[c]

        current.end_word = True


    def search(self, word: str) -> bool:
        current = self.__root

        for c in word:
            children = current.children

            if c not in children:
                return False

            current = children[c]

        return current.end_word

    def startsWith(self, prefix: str) -> bool:
        current = self.__root

        for c in prefix:
            children = current.children

            if c not in children:
                return False

            current = children[c]

        return True
